### PR TITLE
Support while/for-statement in CuPy JIT

### PR DESCRIPTION
**Pull request by asi1024**
Part of #4290.

~Depends on #4672.~

### CODE REVIEW

File: tests/cupy_tests/functional_tests/test_vectorize.py [hunk 1]
```diff
@@ -379,6 +379,79 @@ def func_if(x):
...
+
+        f = xp.vectorize(func_for)
+        x = xp.array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
+        return f(x)
+
+    @testing.numpy_cupy_array_equal()
+    def test_for_range_step(self, xp):
+        def func_for(x, y, z):
+            res = 0
+            for i in range(x, y, z):
```
Comment (emcastillo) [inline tests/cupy_tests/functional_tests/test_vectorize.py:422]: Can we add a test that checks for compound expressions inside range?
```
for i in range(x, some_function()+1):
```

**Review by leofang — COMMENTED** [review]
Question: For the `while` loop, what happens if the condition can only be determined by a device synchronization? (Say, `while x>0:` with `x` being a 0d array.)

**Comment by asi1024** [comment]
@leofang Device synchronization is out of support scope of `cupy.vectorize`. `vectorize` is for element-wise calculation. The input arguments are copied to each local variable at the top of the global function, and the results are written back from the local variable to the global memory at the end of the function.
We will support RawKernel construction from CuPy JIT in the near future, and device synchronization will be supported in it.

**Comment by asi1024** [comment]
Found a bug and fixed it in 6cdb394.

File: tests/cupy_tests/functional_tests/test_vectorize.py (hunk 1 above)
Comment (asi1024) [reply tests/cupy_tests/functional_tests/test_vectorize.py:422]: Added a test: 4124f3d

**Review by emcastillo — APPROVED** [review]

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
CI unrelated

**Pull request merged**